packages/swayze/lib/src/widgets/headers/gestures/header_gesture_detector.dart

```diff
@@ -250,67 +274,251 @@ class _HeaderGestureDetectorState extends State<HeaderGestureDetector> {
+    /// action.
+    setCursorState(SystemMouseCursors.basic);
+
+    Actions.invoke(
+      context,
+      HeaderDragStartIntent(
+        draggingPosition: details.localPosition,
+        headers: selectionRange,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  /// Handles header dragging updates.
+  void handleUpdateDraggingHeader(
+    DragUpdateDetails gestureDetails,
+    _HeaderGestureDetails details,
+  ) {
+    Actions.invoke(
+      context,
+      HeaderDragUpdateIntent(
+        draggingPosition: gestureDetails.localPosition,
+        header: details.headerPosition,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  void handleDragEnd(SwayzeHeaderDragState state) {
+    if (state.isDropAllowed) {
+      Actions.invoke(
+        context,
+        HeaderDragEndIntent(
+          header: state.dropAtIndex,
+          axis: widget.axis,
+        ),
+      );
+    } else {
+      handleDragCancel();
+    }
+  }
+
+  void handleDragCancel() {
+    Actions.invoke(
+      context,
+      HeaderDragCancelIntent(widget.axis),
+    );
+  }
+
+  /// Sets a new cursor state.
+  void setCursorState(MouseCursor newCursor) {
+    if (newCursor != cursor) {
+      setState(() => cursor = newCursor);
+    }
+  }
+
+  /// Checks if a header is selected.
+  bool isHeaderSelected(int position, Axis axis) =>
+      internalScope.controller.selection.isHeaderSelected(position, axis);
+
+  /// Finds the selection that is under position.
+  ///
+  /// Returns null if position header is not selected.
+  HeaderUserSelectionModel? hoverSelection(int position, Axis axis) {
+    final selectionController = internalScope.controller.selection;
+    final selections = selectionController.userSelectionState.selections;
+
+    for (final selection in selections) {
+      if (selection is HeaderUserSelectionModel && selection.axis == axis) {
+        final range = Range(selection.start, selection.end);
+        if (range.contains(position)) {
+          return selection;
+        }
+      }
+    }
+    return null;
+  }
+
+  /// Finds the range of the current reference selection and all adjacent
+  /// selections.
+  ///
+  /// Returns a range containing all adjacent selections from the reference
+  /// selection, so it can be dragged as a single group.
+  Range headerSelectionRange(HeaderUserSelectionModel referenceSelection) {
+    final selectionController = internalScope.controller.selection;
+    final selections = selectionController.userSelectionState.selections
+        .whereType<HeaderUserSelectionModel>();
+
+    var selectionRange =
+        Range(referenceSelection.start, referenceSelection.end);
+
+    final sortedSelections = selections
+        .where((selection) => selection.axis == referenceSelection.axis)
+        .sorted((lhs, rhs) => lhs.start.compareTo(rhs.start));
+
+    final selectionIndex = sortedSelections.indexOf(referenceSelection);
+
+    void checkForAdjacentSelection(HeaderUserSelectionModel selection) {
```

Comment:
Nit: this is now updating the `selectionRange`, so we should probably call it `updateSelection` or `expandSelection`